Error with anaconda/gcc when installing python AMICI on macOS

When trying to install python AMICI on macOS 10.13.2 (with gcc version 4.8.5) I get the following error:

```ruby
  copying amici/swig/solver_idas.i -> build/lib.macosx-10.7-x86_64-3.6/amici/swig
  copying amici/swig/std_unique_ptr.i -> build/lib.macosx-10.7-x86_64-3.6/amici/swig
  running build_clib
  building 'amici' library
  creating build/temp.macosx-10.7-x86_64-3.6
  creating build/temp.macosx-10.7-x86_64-3.6/amici
  creating build/temp.macosx-10.7-x86_64-3.6/amici/src
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/anaconda3/include -arch x86_64 -I/anaconda3/include -arch x86_64 -Iamici/include -Iamici/ThirdParty/SuiteSparse/KLU/Include/ -Iamici/ThirdParty/SuiteSparse/AMD/Include/ -Iamici/ThirdParty/SuiteSparse/COLAMD/Include/ -Iamici/ThirdParty/SuiteSparse/BTF/Include/ -Iamici/ThirdParty/SuiteSparse/SuiteSparse_config/Include/ -Iamici/ThirdParty/SuiteSparse/include -Iamici/ThirdParty/sundials/include -Iamici/ThirdParty/sundials/src -c amici/src/forwardproblem.cpp -o build/temp.macosx-10.7-x86_64-3.6/amici/src/forwardproblem.o -std=c++11
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++ [enabled by default]
  In file included from amici/include/amici/defines.h:3:0,
                   from amici/include/amici/exception.h:4,
                   from amici/include/amici/model.h:4,
                   from amici/src/forwardproblem.cpp:1:
  /anaconda3/envs/my_env/gcc/include/c++/cmath:44:18: fatal error: math.h: No such file or directory
   #include <math.h>
                    ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for amici
```

Does anybody has an idea how to solve this problem? 

update: 
after fixing installation of pkg-config (by re-installing Homebrew) and installing cblas, I have still the following error:

```ruby
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Geometry.h:486:60: error: expected initializer before ‘__WATCHOS_AVAILABLE’
       __OSX_AVAILABLE_STARTING( __MAC_10_12, __IPHONE_10_0 ) __WATCHOS_AVAILABLE(__WATCHOS_3_0) __TVOS_AVAILABLE(__TVOS_10_0);
                                                              ^
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Geometry.h:499:60: error: expected initializer before ‘__WATCHOS_AVAILABLE’
       __OSX_AVAILABLE_STARTING( __MAC_10_12, __IPHONE_10_0 ) __WATCHOS_AVAILABLE(__WATCHOS_3_0) __TVOS_AVAILABLE(__TVOS_10_0);
                                                              ^
  In file included from /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/vImage.h:282:0,
                   from /System/Library/Frameworks/Accelerate.framework/Headers/Accelerate.h:24,
                   from amici/src/cblas.cpp:10:
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Transform.h:385:54: error: expected initializer before ‘__WATCHOS_AVAILABLE’
                                                        __WATCHOS_AVAILABLE(__WATCHOS_2_0);
                                                        ^
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Transform.h:442:54: error: expected initializer before ‘__WATCHOS_AVAILABLE’
                                                        __WATCHOS_AVAILABLE(__WATCHOS_2_0);
                                                        ^
  In file included from /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/vImage.h:282:0,
                   from /System/Library/Frameworks/Accelerate.framework/Headers/Accelerate.h:24,
                   from amici/src/cblas.cpp:10:
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Transform.h:741:60: error: expected initializer before ‘__WATCHOS_AVAILABLE’
       __OSX_AVAILABLE_STARTING( __MAC_10_12, __IPHONE_10_0 ) __WATCHOS_AVAILABLE(__WATCHOS_3_0) __TVOS_AVAILABLE(__TVOS_10_0);
                                                              ^
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Transform.h:749:60: error: expected initializer before ‘__WATCHOS_AVAILABLE’
       __OSX_AVAILABLE_STARTING( __MAC_10_12, __IPHONE_10_0 ) __WATCHOS_AVAILABLE(__WATCHOS_3_0) __TVOS_AVAILABLE(__TVOS_10_0);
                                                              ^
  /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/Transform.h:1047:61: error: expected initializer before ‘__WATCHOS_AVAILABLE’
                                                               __WATCHOS_AVAILABLE(__WATCHOS_2_0);
                                                               ^
  error: command 'gcc' failed with exit status 1
```

Somewhere down the road:
`  /System/Library/Frameworks/vecLib.framework/Headers/vBasicOps.h:377:84: error: cannot convert ‘vUInt16 {aka __vector(8) short unsigned int}’ to ‘__m128i {aka __vector(2) long long int}’ for argument ‘1’ to ‘__m128i _mm_mullo_epi16(__m128i, __m128i)’ `

This is fixed by passing  `-flax-vector-conversions`  to gcc

Furthest we got was:
```
In file included from /usr/include/unistd.h:72:0,
                 from /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/vImage_Types.h:22,
                 from /System/Library/Frameworks/Accelerate.framework/Frameworks/vImage.framework/Headers/vImage.h:258,
                 from /System/Library/Frameworks/Accelerate.framework/Headers/Accelerate.h:24,
                 from /......../src/cblas.cpp:10:
/usr/include/sys/unistd.h:189:72: error: expected initializer before ‘__OSX_AVAILABLE’
 int setattrlistat(int, const char *, void *, void *, size_t, uint32_t) __OSX_AVAILABLE(10.13) __IOS_AVAILABLE(11.0) __TVOS_AVAILABLE(11.0) __WATCHOS_AVAILABLE(4.0);
                                                                        ^
``` 
This seems to be due to some messed up paths leading to the inclusion of an incompatible unistd header.

More fun when using clang instead:
```
  In file included from amici/src/forwardproblem.cpp:1:
  amici/include/amici/model.h:104:22: error: no type named 'unique_ptr' in namespace 'std'
          virtual std::unique_ptr<Solver> getSolver() = 0;
                  ~~~~~^
```
Due to some ancient C++ standard library on this combination of clang/anaconda/osx.

-> `CFLAGS="-stdlib=libc++" CC=clang pip3 install --verbose amici` ?

--- a/python/amici/__init__.py
+++ b/python/amici/__init__.py
@@ -160,11 +160,9 @@ def getFieldAsNumPyArray(rdata, field):
 
     attr = getattr(rdata, field)
     if field in fieldDimensions.keys():
-        if len(fieldDimensions[field]) == 1:
-            return np.array(attr)
-        elif len(attr) == 0:
+        if len(attr) == 0:
             return None
         else:
-            return np.array(attr).reshape(fieldDimensions[field])
+            return stdVec2ndarray(attr, *fieldDimensions[field]).copy()
     else:
         return float(attr)

--- a/python/sdist/MANIFEST.in
+++ b/python/sdist/MANIFEST.in
@@ -1,7 +1,7 @@
 recursive-include amici/ThirdParty/SuiteSparse *.c *.h LICENSE.txt
 recursive-include amici/ThirdParty/sundials *.c *.h LICENSE
 recursive-include amici/include *.h
-recursive-include amici/swig *.i
+recursive-include amici/swig *.i *.h
 include amici/*
 include amici/src/*template*
 include amici/swig/CMakeLists_model.txt

--- a/python/sdist/setup.py
+++ b/python/sdist/setup.py
@@ -22,7 +22,7 @@ import sysconfig
 import pkgconfig
 import subprocess
 from shutil import copyfile
-
+import numpy as np # for include directory
 import setup_clibs  # Must run from within containing directory
 import shutil
 
@@ -116,7 +116,8 @@ amici_module = Extension(
     include_dirs=['amici/include',
                   *libsundials[1]['include_dirs'],
                   *libsuitesparse[1]['include_dirs'],
-                  *h5pkgcfg['include_dirs']
+                  *h5pkgcfg['include_dirs'],
+                  np.get_include()
                   ],
     # Cannot use here, see above
     # libraries=[
@@ -249,7 +250,7 @@ with open("README.md", "r") as fh:
 
 
 def getPackageVersion():
-    return '0.7.1a3'
+    return '0.7.4'
 
 
 # Remove the "-Wstrict-prototypes" compiler option, which isn't valid for

--- a/scripts/buildAmici.sh
+++ b/scripts/buildAmici.sh
@@ -11,15 +11,15 @@ mkdir -p ${AMICI_PATH}/build
 cd ${AMICI_PATH}/build
 CPPUTEST_BUILD_DIR=${AMICI_PATH}/ThirdParty/cpputest-master/build/
 CppUTest_DIR=${CPPUTEST_BUILD_DIR} cmake -DCMAKE_BUILD_TYPE=Debug ..
-make
+make VERBOSE=1
 
 # Disabled until cmake package is made compatible with updated setup.py
 #make python-wheel
 #pip3 install --user --prefix= `ls -t ${AMICI_PATH}/build/python/amici-*.whl | head -1`
 
-make python-sdist
+make VERBOSE=1 python-sdist
 set -x
 python3 -m venv ${AMICI_PATH}/build/venv --clear
 source ${AMICI_PATH}/build/venv/bin/activate
-pip3 install --upgrade pip setuptools pkgconfig wheel
+pip3 install --upgrade pip setuptools pkgconfig wheel numpy
 pip3 install --verbose `ls -t ${AMICI_PATH}/build/python/amici-*.tar.gz | head -1`

--- a/swig/amici.i
+++ b/swig/amici.i
@@ -1,5 +1,25 @@
 %module amici
 
+%{
+#define SWIG_FILE_WITH_INIT
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/arrayobject.h>
+#include <vector>
+#include <stdexcept>
+%}
+%init %{
+import_array();
+%}
+
+%{
+static_assert (sizeof(double) == sizeof (npy_double), "Numpy double size mismatch");
+static_assert (sizeof(int) == sizeof (npy_int), "Numpy integer size mismatch");
+
+#include "../swig/stdvec2numpy.h"
+using namespace amici;
+%}
+%include "stdvec2numpy.h"
+
 typedef double realtype;
 
 %include <stl.i>

--- a/swig/stdvec2numpy.h
+++ b/swig/stdvec2numpy.h
@@ -1,0 +1,128 @@
+namespace amici {
+
+/**
+ * @brief Convert 1D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<double>& vec, int dim1) {
+    if (vec.size() != (unsigned) dim1) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[1] = { dim1 };
+    PyObject * array = PyArray_SimpleNewFromData(1, dims,  NPY_DOUBLE, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 2D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<double>& vec, int dim1, int dim2) {
+    if (vec.size() != (unsigned) dim1 * dim2) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[2] = { dim1, dim2 };
+    PyObject * array = PyArray_SimpleNewFromData(2, dims,  NPY_DOUBLE, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 3D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @param dim3
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<double>& vec, int dim1, int dim2, int dim3) {
+    if (vec.size() != (unsigned) dim1 * dim2 * dim3) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[3] = { dim1, dim2, dim3 };
+    PyObject * array = PyArray_SimpleNewFromData(3, dims,  NPY_DOUBLE, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 2D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @param dim3
+ * @param dim4
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<double>& vec, int dim1, int dim2, int dim3, int dim4) {
+    if (vec.size() != (unsigned) dim1 * dim2 * dim3 * dim4) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[4] = { dim1, dim2, dim3, dim4 };
+    PyObject * array = PyArray_SimpleNewFromData(4, dims,  NPY_DOUBLE, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+
+/**
+ * @brief Convert 1D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<int>& vec, int dim1) {
+    if (vec.size() != (unsigned) dim1) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[1] = { dim1 };
+    PyObject * array = PyArray_SimpleNewFromData(1, dims,  NPY_INT, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 2D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<int>& vec, int dim1, int dim2) {
+    if (vec.size() != (unsigned) dim1 * dim2) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[2] = { dim1, dim2 };
+    PyObject * array = PyArray_SimpleNewFromData(2, dims,  NPY_INT, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 3D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @param dim3
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<int>& vec, int dim1, int dim2, int dim3) {
+    if (vec.size() != (unsigned) dim1 * dim2 * dim3) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[3] = { dim1, dim2, dim3 };
+    PyObject * array = PyArray_SimpleNewFromData(3, dims,  NPY_INT, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+/**
+ * @brief Convert row-major flattened 2D array to *non-owning* numpy ndarray.
+ * @param vec
+ * @param dim1
+ * @param dim2
+ * @param dim3
+ * @param dim4
+ * @return
+ */
+PyObject* stdVec2ndarray(std::vector<int>& vec, int dim1, int dim2, int dim3, int dim4) {
+    if (vec.size() != (unsigned) dim1 * dim2 * dim3 * dim4) throw std::runtime_error("Size mismatch in stdVec2ndarray");
+    npy_intp dims[4] = { dim1, dim2, dim3, dim4 };
+    PyObject * array = PyArray_SimpleNewFromData(4, dims,  NPY_INT, vec.data());
+    if (!array) throw std::runtime_error("Unknown failure in stdVec2ndarray");;
+    return array;
+}
+
+}

